### PR TITLE
sql: parse PG CONNECTION DETAILS + PUBLICATION as options

### PR DESCRIPTION
**Pull request by sploiselle**
@benesch mentioned this change in https://github.com/MaterializeInc/materialize/pull/14539#discussion_r961104273, so here it is

### Motivation

This PR adds a known-desirable feature.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - PostgreSQL sources must now specify `PUBLICATION` an an option in parentheses, e.g. `CREATE SOURCE...POSTGRES CONNECTION (PUBLICATION 'x')...`

### CODE REVIEW

**Comment by benesch** [comment]
Lovely! Unfortunately this needs to wait until our next backwards-incompatible release, which is currently scheduled for 9/27. Unless you wanted to write a migration—but doesn't seem worth it since it'll be a ton of complexity.

**Comment by sploiselle** [comment]
@benesch nbd to hold off and rebase this

File: src/sql/src/plan/statement/ddl.rs [hunk 1]
```diff
@@ -613,7 +620,8 @@ pub fn plan_create_source(
                 ProtoPostgresSourceDetails::decode(&*details).map_err(|e| sql_err!("{}", e))?;
             let connection = SourceConnection::Postgres(PostgresSourceConnection {
                 connection,
-                publication: publication.clone(),
+                publication: publication
+                    .ok_or_else(|| sql_err!("POSTGRESS CONNECTION must specify PUBLICATION"))?,
```
Comment (benesch) [inline src/sql/src/plan/statement/ddl.rs:624]:
```suggestion
                    .ok_or_else(|| sql_err!("POSTGRES CONNECTION must specify PUBLICATION"))?,
```

File: src/sql/src/pure.rs [hunk 2]
```diff
@@ -207,20 +207,31 @@ pub async fn purify_create_source(
                     _ => bail!("{} is not a postgres connection", item.name()),
                 }
             };
+            let crate::plan::statement::PgConfigOptionExtracted { publication, .. } =
+                options.clone().try_into()?;
+            let publication = publication.expect("verified existed");
```
Comment (benesch) [inline src/sql/src/pure.rs:212]: Is this true? Is this going to cause the same bug that we just fixed with Kafka connections not specifying their `TOPIC`?

**Review by benesch — APPROVED** [review]
:lgtm_strong:

**Pull request merged**